reco_utils/recommender/newsrec/models/lstur.py

```diff
@@ -0,0 +1,160 @@
+from reco_utils.recommender.newsrec.models.layers import AttLayer2, ComputeMasking, OverwriteMasking
+
+__all__ = ["LSTURModel"]
+
+class LSTURModel(BaseModel):
+    '''LSTUR model(Neural News Recommendation with Multi-Head Self-Attention)
+
+    Mingxiao An, Fangzhao Wu, Chuhan Wu, Kun Zhang, Zheng Liu and Xing Xie: 
+    Neural News Recommendation with Long- and Short-term User Representations, ACL 2019
+
+    Attributes:
+        word2vec_embedding (numpy.array): Pretrained word embedding matrix.
+        hparam (obj): Global hyper-parameters.
+    '''
+    def __init__(self, hparams, iterator_creator_train, iterator_creator_test):
+        """Initialization steps for LSTUR.
+        Compared with the BaseModel, LSTUR need word embedding.
+        After creating word embedding matrix, BaseModel's __init__ method will be called.
+        
+        Args:
+            hparams (obj): Global hyper-parameters. Some key setttings such as type and gru_unit are there.
+            iterator_creator_train(obj): LSTUR data loader class for train data.
+            iterator_creator_test(obj): LSTUR data loader class for test and validation data
+        """
+        
+        self.word2vec_embedding = self._init_embedding(hparams.wordEmb_file)
+        self.hparam = hparams
+        
+        super().__init__(hparams, iterator_creator_train, iterator_creator_test)
+
+    def _init_embedding(self, file_path):
+        """Load pre-trained embeddings as a constant tensor.
+        
+        Args:
+            file_path (str): the pre-trained embeddings filename.
+
+        Returns:
+            np.array: A constant numpy array.
+        """
+        return np.load(file_path).astype(np.float32)
+
+
+    def _build_graph(self):
+        """Build LSTUR model and scorer.
+
+        Returns:
+            obj: a model used to train.
+            obj: a model used to evaluate and inference.
+        """
+
+        model, scorer = self._build_lstur()
+        return model, scorer
+
+    def _build_userencoder(self, titleencoder, type='ini'):
+        '''The main function to create user encoder of LSTUR.
+
+        Args:
+            titleencoder(obj): the news encoder of LSTUR. 
+
+        Return:
+            obj: the user encoder of LSTUR.
+        '''
+        hparams = self.hparams
+        his_input_title = keras.Input(shape=(hparams.his_size, hparams.doc_size),dtype='int32')
+        user_indexes = keras.Input(shape=(1,), dtype='int32')
+
+        user_embedding_layer = layers.Embedding(hparams.user_num,
+                                 hparams.gru_unit,
+                                 trainable=True,
+                                 embeddings_initializer='zeros')
+        
+        long_u_emb = layers.Reshape((hparams.gru_unit,))(user_embedding_layer(user_indexes))
+        click_title_presents = layers.TimeDistributed(titleencoder)(his_input_title)
+
+        if type=='ini':
+            user_present = layers.GRU(hparams.gru_unit,)(
```

Comment:
I have a curiosity, in many occasions I've seen that you have used GRU instead of LSTM. Could you please explain the reason?

Reply:
GRU is faster and smaller than LSTM. And in our experiment, GRU is as effective as LSTM.